feat!: burst sampling added to the resampling updater

ramyamounir opened the pull request:
Before this PR, hypotheses were added in every step with a constant parameter that decides how much to add. This PR introduces a more adaptive approach to sampling that depends on the model's current prediction error. Here are some important changes:
- The distinction between initializing a hypothesis space and sampling has been removed from the code and unit tests. The initial sampling is now similar to sampling elsewhere in the episode, moving us more towards episode-free experiments.
- The sampling happens in bursts. Each burst is activated when the highest global (i.e., over all objects and channels) slope is below some threshold. The burst happens over a few consecutive steps to smooth out the effect of sensor noise. This burst happens also at the beginning of an episode. 
- Since the highest slope calculation is global over all object ids, it has to happen **once and before** the multiple calls to `update_hypotheses`. This led to declaring `pre_step` and `post_step` functions in all updaters (protocol, `DefaultHypothesesUpdater`, `ResamplingHypothesesUpdater`).

<img width="1722" height="886" alt="burst" src="https://github.com/user-attachments/assets/2583feeb-8b96-4758-b8f7-c3d79c0e7971" />


## Benchmarks

Before running full benchmark experiments, we will be running hyper-parameter tuning to find out good defaults for the burst sampling approach. Here are the current defaults and the expected variations to be tested.
```python
        resampling_multiplier: float = 0.2,  #[0.1, 0.2, 0.3, 0.4]
        evidence_slope_threshold: float = 0.5, #[0.4, 0.5, 0.6]
        sampling_burst_duration: int = 5,
        slope_burst_trigger: float = 1.0, #[0.9,1.0,1.1]
```
Here are some results for the randrot noise 77 objects experiments with surface and distant agents. It's a 3x3 grid where each row shows results for varying a different parameter (resampling_multiplier, evidence_slope_threshold, slope_burst_trigger), and each column shows a different metric (percent_correct, rotation_error, runtime). If a parameter is not varied, the default parameter above is used. 

**Runs shown below are on WandB tagged under `feat.dynamic_resizing:PR#6` with experiment names starting with `burst_`**

<img width="1722" height="886" alt="params" src="https://github.com/user-attachments/assets/391d3738-2ba9-4430-a15b-b2ef04cd88a7" />

ramyamounir committed:
feat!: add intelligent sampling

feat! add intelligent sampling based on highest slope

ramyamounir committed:
tests: remove initial count tests and init_hyp_space boolean

ramyamounir commented:
@vkakerbeck, feel free to start reviewing if you like. I will be adding parameter tuning results soon.

ramyamounir commented:
@vkakerbeck, I've added some bar plots in the PR description. Tagging you here because I don't think it notifies you when I edit the PR description.

ramyamounir commented:
Based on the parameter bar plots, I ran the below values for all of the benchmarks. We seem to have improved performance at the same or slightly improved runtime, which is great.

```python
        resampling_multiplier: float = 0.4,
        evidence_slope_threshold: float = 0.5,
        sampling_burst_duration: int = 5,
        slope_burst_trigger: float = 1.0,
```

NOTE: I don't know what's going on with the `base_10multi_distinctobj_dist_agent`, @nielsleadholm had some comments about why this is expected for this experiment, but I don't remember what the issue was. I can look into it later.

**Runs are on WandB tagged under `feat.dynamic_resizing:PR#6`.**

### base_config_10distinctobj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 0 | 0.71 | 0.71 | ❌ |
| match_steps | 34 | 37 | 3 | ❌ |
| rotation_error (deg) | 13.31 | 9.44 | -3.87 | ✅ |
| runtime (min) | 2 | 2 | 0 | ⬜ |
| episode_runtime (sec) | 8 | 9 | 1 | ❌ |

### base_config_10distinctobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 0 | 0.71 | 0.71 | ❌ |
| match_steps | 26 | 28 | 2 | ❌ |
| rotation_error (deg) | 15.03 | 6.22 | -8.81 | ✅ |
| runtime (min) | 3 | 3 | 0 | ⬜ |
| episode_runtime (sec) | 11 | 12 | 1 | ❌ |

### randrot_noise_10distinctobj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 3 | 0 | -3 | ✅ |
| match_steps | 52 | 38 | -14 | ✅ |
| rotation_error (deg) | 22.41 | 19.64 | -2.77 | ✅ |
| runtime (min) | 3 | 2 | -1 | ✅ |
| episode_runtime (sec) | 20 | 18 | -2 | ✅ |

### randrot_noise_10distinctobj_dist_on_distm

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 99 | 100 | 1 | ✅ |
| used_mlh (%) | 1 | 0 | -1 | ✅ |
| match_steps | 37 | 32 | -5 | ✅ |
| rotation_error (deg) | 12.27 | 14.6 | 2.33 | ❌ |
| runtime (min) | 3 | 2 | -1 | ✅ |
| episode_runtime (sec) | 17 | 17 | 0 | ⬜ |

### randrot_noise_10distinctobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 1 | 0 | -1 | ✅ |
| match_steps | 29 | 30 | 1 | ❌ |
| rotation_error (deg) | 20.12 | 8.82 | -11.3 | ✅ |
| runtime (min) | 3 | 3 | 0 | ⬜ |
| episode_runtime (sec) | 21 | 24 | 3 | ❌ |

### randrot_10distinctobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 0 | 0 | 0 | ⬜ |
| match_steps | 28 | 31 | 3 | ❌ |
| rotation_error (deg) | 19.36 | 5.54 | -13.82 | ✅ |
| runtime (min) | 2 | 2 | 0 | ⬜ |
| episode_runtime (sec) | 11 | 13 | 2 | ❌ |

### randrot_noise_10distinctobj_5lms_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 100 | 100 | 0 | ⬜ |
| used_mlh (%) | 0 | 0 | 0 | ⬜ |
| match_steps | 52 | 55 | 3 | ❌ |
| rotation_error (deg) | 44.71 | 41.38 | -3.33 | ✅ |
| runtime (min) | 5 | 5 | 0 | ⬜ |
| episode_runtime (sec) | 37 | 40 | 3 | ❌ |

### base_10simobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 93.57 | 100 | 6.43 | ✅ |
| used_mlh (%) | 8.57 | 6.43 | -2.14 | ✅ |
| match_steps | 70 | 61 | -9 | ✅ |
| rotation_error (deg) | 12.34 | 3.04 | -9.3 | ✅ |
| runtime (min) | 6 | 4 | -2 | ✅ |
| episode_runtime (sec) | 27 | 21 | -6 | ✅ |

### randrot_noise_10simobj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 84 | 89 | 5 | ✅ |
| used_mlh (%) | 40 | 21 | -19 | ✅ |
| match_steps | 234 | 173 | -61 | ✅ |
| rotation_error (deg) | 31.08 | 25.14 | -5.94 | ✅ |
| runtime (min) | 13 | 10 | -3 | ✅ |
| episode_runtime (sec) | 107 | 80 | -27 | ✅ |

### randrot_noise_10simobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 93 | 100 | 7 | ✅ |
| used_mlh (%) | 34 | 30 | -4 | ✅ |
| match_steps | 172 | 173 | 1 | ❌ |
| rotation_error (deg) | 22.81 | 16.8 | -6.01 | ✅ |
| runtime (min) | 17 | 18 | 1 | ❌ |
| episode_runtime (sec) | 141 | 150 | 9 | ❌ |

### randomrot_rawnoise_10distinctobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 67 | 76 | 9 | ✅ |
| used_mlh (%) | 77 | 70 | -7 | ✅ |
| match_steps | 15 | 16 | 1 | ❌ |
| rotation_error (deg) | 117.12 | 101.89 | -15.23 | ✅ |
| runtime (min) | 4 | 4 | 0 | ⬜ |
| episode_runtime (sec) | 6 | 9 | 3 | ❌ |

### base_10multi_distinctobj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 80.71 | 0.71 | -80 | ❌ |
| used_mlh (%) | 14.29 | 0 | -14.29 | ✅ |
| match_steps | 30 | 21 | -9 | ✅ |
| rotation_error (deg) | 19.6 | 1.22 | -18.38 | ✅ |
| runtime (min) | 45 | 1 | -44 | ✅ |
| episode_runtime (sec) | 1 | 1 | 0 | ⬜ |

### base_77obj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 93.51 | 95.24 | 1.73 | ✅ |
| used_mlh (%) | 12.99 | 5.63 | -7.36 | ✅ |
| match_steps | 107 | 80 | -27 | ✅ |
| rotation_error (deg) | 16.46 | 13.54 | -2.92 | ✅ |
| runtime (min) | 20 | 14 | -6 | ✅ |
| episode_runtime (sec) | 65 | 48 | -17 | ✅ |

### base_77obj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 98.7 | 100 | 1.3 | ✅ |
| used_mlh (%) | 6.49 | 2.6 | -3.89 | ✅ |
| match_steps | 56 | 41 | -15 | ✅ |
| rotation_error (deg) | 11.38 | 4.29 | -7.09 | ✅ |
| runtime (min) | 13 | 10 | -3 | ✅ |
| episode_runtime (sec) | 38 | 31 | -7 | ✅ |

### randrot_noise_77obj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 89.18 | 93.07 | 3.89 | ✅ |
| used_mlh (%) | 22.08 | 15.15 | -6.93 | ✅ |
| match_steps | 147 | 131 | -16 | ✅ |
| rotation_error (deg) | 37.37 | 29.3 | -8.07 | ✅ |
| runtime (min) | 33 | 34 | 1 | ❌ |
| episode_runtime (sec) | 120 | 123 | 3 | ❌ |

### randrot_noise_77obj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 93.94 | 100 | 6.06 | ✅ |
| used_mlh (%) | 23.81 | 19.48 | -4.33 | ✅ |
| match_steps | 115 | 107 | -8 | ✅ |
| rotation_error (deg) | 36.7 | 26.29 | -10.41 | ✅ |
| runtime (min) | 31 | 32 | 1 | ❌ |
| episode_runtime (sec) | 111 | 116 | 5 | ❌ |

### randrot_noise_77obj_5lms_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 87.01 | 93.51 | 6.5 | ✅ |
| used_mlh (%) | 0 | 0 | 0 | ⬜ |
| match_steps | 64 | 67 | 3 | ❌ |
| rotation_error (deg) | 60.61 | 49.59 | -11.02 | ✅ |
| runtime (min) | 9 | 13 | 4 | ❌ |
| episode_runtime (sec) | 86 | 127 | 41 | ❌ |

### unsupervised_inference_distinctobj_dist_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 98 | 97 | -1 | ❌ |
| match_steps | 99 | 99 | 0 | ⬜ |
| runtime (min) | 29 | 11 | -18 | ✅ |
| episode_runtime (sec) | 17 | 6 | -11 | ✅ |

### unsupervised_inference_distinctobj_surf_agent

| Metric | Baseline | Proposed | Δ | Result |
| --- | --- | --- | --- | --- |
| percent_correct (%) | 95 | 100 | 5 | ✅ |
| match_steps | 100 | 99 | -1 | ✅ |
| runtime (min) | 22 | 16 | -6 | ✅ |
| episode_runtime (sec) | 13 | 10 | -3 | ✅ |

ramyamounir committed:
chore: update resampling multiplier parameter to 0.4

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -73,23 +73,41 @@ class ChannelHypothesesResamplingTelemetry(ChannelHypothesesUpdateTelemetry):
+    and rebuilding the hypothesis space when the model's prediction error is high. The
+    prediction error is determined based on the highest evidence slope over all the
+    objects hypothesis spaces. If the hypothesis with the highest slope is unable to
+    accumulate evidence at a high enough slope, a sampling burst is triggered. A
```
Maybe to make this a bit more clear, add something like `i.e. none of the current hypotheses match the incoming observations well`

ramyamounir replied:
Updated in [94bb6c2af494](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/94bb6c2af4940eb5d0262caef5690696a477534a)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -73,23 +73,41 @@ class ChannelHypothesesResamplingTelemetry(ChannelHypothesesUpdateTelemetry):
+    objects hypothesis spaces. If the hypothesis with the highest slope is unable to
+    accumulate evidence at a high enough slope, a sampling burst is triggered. A
+    sampling burst adds new hypotheses over a specified `sampling_burst_duration`
+    number of consecutive steps to all hypothesis spaces. This burst duration smoothes
```
```suggestion
    number of consecutive steps to all hypothesis spaces. This burst duration reduces
```

ramyamounir replied:
Updated in [94bb6c2af494](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/94bb6c2af4940eb5d0262caef5690696a477534a)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -73,23 +73,41 @@ class ChannelHypothesesResamplingTelemetry(ChannelHypothesesUpdateTelemetry):
-    hypothesis space, as well as new hypotheses informed by the sensed pose.
-
-    The resampling process is governed by two main parameters:
+    """Hypotheses updater that resamples hypotheses based on prediction error.
```
```suggestion
    """Hypotheses updater that adds and deletes hypotheses based on evidence slope.
```

ramyamounir replied:
Updated in [94bb6c2af494](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/94bb6c2af4940eb5d0262caef5690696a477534a)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -73,23 +73,41 @@ class ChannelHypothesesResamplingTelemetry(ChannelHypothesesUpdateTelemetry):
         as a multiplier of the object graph nodes.
       - `evidence_slope_threshold`: Hypotheses below this threshold are deleted.
+      - `sampling_burst_duration`: The number of consecutive steps in each burst.
+      - `slope_burst_trigger`: The threshold for triggering a sampling burst.
```
it would be good to elaborate a bit more here. What value is being thresholded and what range should this value be in?

ramyamounir replied:
Updated in [94bb6c2af494](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/94bb6c2af4940eb5d0262caef5690696a477534a)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -73,23 +73,41 @@ class ChannelHypothesesResamplingTelemetry(ChannelHypothesesUpdateTelemetry):
+    Note that if the PC of the first observation is undetermined,
+    `resampling_multiplier` should be set to `umbilical_num_poses` to reproduce the
+    exact results of `DefaultHypothesesUpdater`. In practice, this is difficult to
+    predict because it relies on the first sampled observation.
```
As discussed yesterday, I think it would be worth testing this update to how `resampling_multiplier` is used:
- If you want to sample hypotheses for all locations on the object model, set `resampling_multiplier=1`. This is the max value for this parameter and usually you would want to set it to a fraction of the number of nodes in the graph.
- for each location on the graph you sample 2 pose hypotheses by default and `umbilical_num_poses` if the sensed curvature directions are undefined.
- This means you effectively add `num_graph_nodes * resampling_multiplier * 2` new hypotheses if the curvature is defined and `num_graph_nodes * resampling_multiplier * umbilical_num_poses` new hypotheses if the curvature is undefined.

This would mean that on some steps you add more hypotheses than on others and you may add slightly more than you do in the current implementation. However, I think it is reasonable to add more if curvature is undefined + your parameter experiments look like it overall has a small effect on runtime to add more hypotheses.

ramyamounir replied:
This makes sense, but I want to think about it a bit more because now that we're not fully dependent on a single first observation, I don't know if it will be beneficial to sample more hypotheses on umbilical points. I'm speculating that sampling the same amount of hypotheses and moving on to the next observation will be empirically better. If we don't get good enough hypotheses from these 2 rotations, we will keep bursting until we do.

If the object is a sphere or a plate, then we will always get umbilical points and it won't matter if we sample 2 or 8 because it's symmetric. But if it's not, then we will soon get PC defined points and the excessive sampling we did when the PC was undefined will not be very effective anyway. Uniformly sampling 8 rotations seems like it will accumulate too much error as the sensor moves a few steps.

I think this is one of those things that can be tested empirically. Maybe we can chat about this more tomorrow.

vkakerbeck replied:
Sounds reasonable, especially as we add more sophisticated policies that would move us to more interesting locations quickly.
Just to make sure I understand everything you are saying, what do you mean by "Uniformly sampling 8 rotations seems like it will accumulate too much error as the sensor moves a few steps."?

ramyamounir replied:
> what do you mean by "Uniformly sampling 8 rotations seems like it will accumulate too much error as the sensor moves a few steps."?

When the PC is undefined, we are randomly sampling 8 rotations on the surface normal plane hoping that one of them will be "close" to the correct orientation of the object. I think 8 is too coarse and it's unlikely to get a close enough hypothesis that keeps its slope above `deletion_trigger_slope`. All 8 rotations will probably be deleted in a few steps because we are aggressively deleting hypotheses based on their slopes. I also think the problem gets worse over time as the sensor moves a few more steps; a small pose error might be okay in the vicinity of the sampled observation, but will definitely accumulate more error as we move away after a few steps. 

I think this approach made more sense when we weren't deleting and adding hypotheses, so monty would just try to choose the best possible hypothesis even if it means higher pose error. Now this best possible hypotheses would be deleted. I doubt any of the 8 uniformly sampled rotations will survive the deletion threshold, and will only make runtime a bit worse since we have to keep them for 5 steps.

vkakerbeck replied:
Hm so if we sample 8 rotations uniformly and assume the point normal is correct, wouldn't the maximum possible pose error be ~22 degrees? Is this still too large to survive deletion?

ramyamounir replied:
Yeah I think so. It really depends on how irregular the object surface is and how far the sensor has moved from the location where it sampled this hypothesis. I've seen it delete good hypotheses with low pose errors because of some noise. But maybe we can determine this empirically if necessary.

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -106,8 +124,10 @@ def __init__(
+        resampling_multiplier: float = 0.4,
+        evidence_slope_threshold: float = 0.5,
+        sampling_burst_duration: int = 5,
+        slope_burst_trigger: float = 1.0,
```
these are nit picks about naming but `evidence_slope_threshold` and `slope_burst_trigger` both are thresholds on the slope but one variable name expresses this while the other tells you what it is used for. How about `deletion_trigger_slope` and `burst_trigger_slope`?

ramyamounir replied:
Yeah, that's better. Updated in [f2c0d7b6aff7](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/f2c0d7b6aff77d71f637669e8b0fd876bc1c1ca7)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -144,10 +164,14 @@ def __init__(
                 as a multiplier of the object graph nodes. Value of 0.0 results in no
                 resampling. Value can be greater than 1 but not to exceed the
-                `num_hyps_per_node` of the current step. Defaults to 0.1.
+                `num_hyps_per_node` of the current step. Defaults to 0.2.
```
```suggestion
                `num_hyps_per_node` of the current step. Defaults to 0.4.
```

ramyamounir replied:
Nice catch, updated in [94bb6c2af494](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/94bb6c2af4940eb5d0262caef5690696a477534a)

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -222,6 +248,31 @@ def __init__(
+        ):
+            self.sampling_burst_steps = self.sampling_burst_duration
+
+    def post_step(self) -> None:
```
Do we need to do this in a separate post_step method? Couldn't we just decremen it in the `if self.sampling_burst_steps > 0:` branch in the `sample_count` function?

ramyamounir replied:
Yes, we have to decrement this value only a single time every step. It counts how many steps we have left in the burst. If we do it in the `sample_count` function, it will be decremented every object hypotheses space (i.e., more than one time per step).

vkakerbeck replied:
ok thanks, that makes sense

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -730,3 +777,26 @@ def remap_hypotheses_ids_to_present(
+        max_slope = float("-inf")
+
+        for tr in self.evidence_slope_trackers.values():
+            for ch in tr.evidence_buffer.keys():
```
another nit: I am not a fan of variable names like tr and ch. Why not just call it tracker and channel?

ramyamounir replied:
Updated in [b405c04bc626](https://github.com/thousandbrainsproject/feat.dynamic_resizing/commit/b405c04bc626ea48ddb82fd8579949cc3bcad26c)

vkakerbeck commented on tests/unit/frameworks/models/evidence_matching/resampling_hypotheses_updater_test.py:
```diff
@@ -160,30 +159,6 @@ def run_sample_count(
-            init_hyp_space=init_hyp_space,
-        )
-
-    def _initial_count(self, rlm, pose_defined):
-        """This tests that the initial requested number of hypotheses is correct.
-
-        In order to initialize a hypothesis space, the `_sample_count` should request
-        that all resampled hypotheses be of the type informed. This tests the informed
-        sampling with defined and undefined poses.
-        """
```
Couldn't you still test that the requested count on the first step is correct given the resampling multiplier? Or is this redundant now since it is already tested elsewhere?

ramyamounir replied:
Yeah exactly, it's redundant. There is no notion of initial count anymore, we just need to test that it's sampling correctly since the same sampling logic is being used everywhere.

vkakerbeck reviewed (commented):
Nice! Just a few smaller comments

ramyamounir committed:
docs: updated wording in docstrings

ramyamounir committed:
refactor: rename tr -> tracker and ch -> channel

ramyamounir committed:
refactor!: updated variable names

evidence_slope_threshold -> deletion_trigger_slope
slope_burst_trigger -> burst_trigger_slope

ramyamounir commented:
Thanks for your detailed review @vkakerbeck! I've addressed your comments except for a more involved one which I'd like to discuss a bit more tomorrow first.

vkakerbeck reviewed (approved):
Great, thanks for the quick updates! All looks good to me :) I'm excited about this change!

ramyamounir committed:
Merge branch 'dev' into burst_sampling
